src/sage/rings/polynomial/multi_polynomial_libsingular.pyx

```diff
@@ -4485,31 +4565,31 @@ cdef class MPolynomial_libsingular(MPolynomial_libsingular_base):
             except Exception:
                 raise NotImplementedError("Factorization of multivariate polynomials over %s is not implemented."%self._parent._base)
 
-        if n_GetChar(_ring.cf) > 1<<29:
-            raise NotImplementedError("Factorization of multivariate polynomials over prime fields with characteristic > 2^29 is not implemented.")
-
-        # I make a temporary copy of the poly in self because singclap_factorize appears to modify it's parameter
-        ptemp = p_Copy(self._poly, _ring)
         iv = NULL
-        sig_on()
         if _ring != currRing:
             rChangeCurrRing(_ring)   # singclap_factorize
```

Comment:
This looks like a repetition of what's done in line 4553.  I realise that `currRing` may be changed by singular, but I don't see where it could have changed here - or does `p_IsConstant` set `currRing`?.  Also, I do not understand the comment `# singclap_factorize`.

Reply:
If I understood correctly the point is you may work with multiple rings at the same time in SageMath e.g.

```
R.<x, y> = ZZ[]
S.<z, t> = GF(5)[]
(x+y).factor()
(z+t).factor()
```

and singular state is sort of global, so you need to swap them around like that.

Not sure (shouldn't each object keep track of what's its ring?), but it's not my code either.

If I understood correctly the comment is "this setting is for the function below", but that seems quite obvious i.e. unnecessary.

p/s. Can you not push to the branch directly? Working with the patches on GitHub interface is a bit cumbersome.

Reply:
Yes, the singular ring is global.

> If I understood correctly the comment is "this setting is for the function below", but that seems quite obvious i.e. unnecessary.

Yes, that makes sense.  Please leave the comment, it seems that this is consistently done.

Unless of course you can determine that setting the ring here is unnecessary.

(I thought you were a singular expert?)

Reply:
No, I just learn enough Singular for this one pull request… possibly dig into the source code if needed. And in the meantime notice several bugs/inefficiencies with the existing code e.g. we could use `perfect_power()` instead of `factor()`, etc.

(on an unrelated note, it looks like FLINT's multivariate polynomial library is more well-developed than Singular though.)